hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/viewfs/TestViewFsOverloadSchemeWithHdfsScheme.java

```diff
@@ -0,0 +1,351 @@
+    final Path hdfsTargetPath = new Path(defaultFSURI + HDFS_USER_FOLDER);
+    final Path localTragetPath = new Path(localTargetDir.toURI());
+
+    createLinks(false, hdfsTargetPath, localTragetPath);
+
+    ViewFsOverloadScheme fs = (ViewFsOverloadScheme) FileSystem.get(conf);
+    try {
+      fs.create(new Path("/newFileOnRoot"));
+      Assert.fail("It should fail as root is read only in viewFS.");
+    } finally {
+      fs.close();
+    }
+  }
+
+  /**
+   * Create mount links as follows
+   * hdfs://localhost:xxx/HDFSUser --> hdfs://localhost:xxx/HDFSUser/
+   * hdfs://localhost:xxx/local --> file://TEST_ROOT_DIR/root/
+   * fallback --> hdfs://localhost:xxx/HDFSUser/
+   *
+   * It will find fallback link, but root is not accessible and read only.
+   */
+  @Test(expected = AccessControlException.class, timeout = 30000)
+  public void testCreateOnRootShouldFailEvenFallBackMountLinkConfigured()
+      throws Exception {
+    final Path hdfsTargetPath = new Path(defaultFSURI + HDFS_USER_FOLDER);
+    final Path localTragetPath = new Path(localTargetDir.toURI());
+
+    createLinks(true, hdfsTargetPath, localTragetPath);
+
+    ViewFsOverloadScheme fs = (ViewFsOverloadScheme) FileSystem.get(conf);
+    try {
+      fs.create(new Path("/onRootWhenFallBack"));
+      Assert.fail(
+          "It should fail as root is read only in viewFS, even when configured"
+              + " with fallback.");
+    } finally {
+      fs.close();
+    }
+  }
+
+  /**
+   * Create mount links as follows
+   * hdfs://localhost:xxx/HDFSUser --> hdfs://localhost:xxx/HDFSUser/
+   * hdfs://localhost:xxx/local --> file://TEST_ROOT_DIR/root/
+   * fallback --> hdfs://localhost:xxx/HDFSUser/
+   *
+   * It will find fallback link, but root is not accessible and read only.
+   */
+  @Test(expected = UnsupportedFileSystemException.class, timeout = 30000)
+  public void testInvalidOverloadSchemeTargetFS() throws Exception {
+    final Path hdfsTargetPath = new Path(defaultFSURI + HDFS_USER_FOLDER);
+    final Path localTragetPath = new Path(localTargetDir.toURI());
+    conf = new Configuration();
+    createLinks(true, hdfsTargetPath, localTragetPath);
+    conf.set(CommonConfigurationKeys.FS_DEFAULT_NAME_KEY,
+        defaultFSURI.toString());
+    conf.set(String.format(FS_IMPL_PATTERN_KEY, HDFS_SCHEME),
+        ViewFsOverloadScheme.class.getName());
+
+    ViewFsOverloadScheme fs = (ViewFsOverloadScheme) FileSystem.get(conf);
+    try {
+      fs.create(new Path("/onRootWhenFallBack"));
+      Assert.fail("OverloadScheme target fs should be valid.");
+    } finally {
+      fs.close();
+    }
+  }
+
+  /**
+   * Create mount links as follows
+   * hdfs://localhost:xxx/HDFSUser --> hdfs://localhost:xxx/HDFSUser/
+   * hdfs://localhost:xxx/local --> file://TEST_ROOT_DIR/root/
+   *
+   * It should be able to create file using ViewFsOverloadScheme.
+   */
+  @Test(timeout = 30000)
+  public void testViewFsOverloadSchemeWhenInnerCacheDisabled()
```

Comment:
Can we add a test that cache actually works?

Reply:
We have not changed cache behavior itself. When cache disabled, we have varied path. 
Anyway Its should be easy to add a test for that I think. Let me add one.

Reply:
Added Tests.